File: Assets/Scripts/Connection/WebSocket/WebSocketServer.cs
using UnityEngine;
using Jering.Javascript.NodeJS;
using Microsoft.Extensions.DependencyInjection;
using UnityEngine.Assertions;
using System.IO;
using ISBEP.Utility;


namespace ISBEP.Communication
{
    public class WebSocketServer : MonoBehaviour
    {
        [Tooltip("Specify whether debug message for the web socket server should be displayed in the logs.")]
        public bool DebugMessages = false;
        private readonly string CONTEXT = "Web Socket Server";

        [Tooltip("Specify if node js debugger should run. See https://nodejs.org/en/learn/getting-started/debugging for debugging information.")]
        public bool nodeJSDebug = false;

        private INodeJSService nodeJSService;
        public static WebSocketServer Instance { get; private set; }

        private void Awake()
        {
            if (DebugMessages) Util.AddDebugContext(CONTEXT);

            // Set static reference to self, so class has public access point.
            Instance = this;

            // Configure NodeJs access path, so we can call the web socket server javascript file.
            string FilePathNodeJS = Path.Join(Application.dataPath, "/Scripts/Connection/WebSocket/NodeJS");

            Util.DebugLog(CONTEXT, $"FilePathNodeJS data path {FilePathNodeJS}");

            // Setup the NodeJS invoke services. With or without debug settings.
            var services = new ServiceCollection();
            services.AddNodeJS();
            if (nodeJSDebug)
            {
                services.Configure<NodeJSProcessOptions>(options =>
                {
                    options.ProjectPath = FilePathNodeJS;
                    options.NodeAndV8Options = "--inspect-brk";
                });
                services.Configure<OutOfProcessNodeJSServiceOptions>(options =>
                {
                    options.InvocationTimeoutMS = -1;
                });
            } else
            {
                services.Configure<NodeJSProcessOptions>(options =>
                {
                    options.ProjectPath = FilePathNodeJS;
                });
            }
            ServiceProvider serviceProvider = services.BuildServiceProvider();
            nodeJSService = serviceProvider.GetRequiredService<INodeJSService>();
        }

        public class Result
        {
            public string Message { get; set; }
        }

        private void Start()
        {
            StartServerWithNodeJS();
        }

        private void OnDisable()
        {
            StopServerWithNodeJS();
        }

        private async void StartServerWithNodeJS()
        {
            Result result = await nodeJSService.InvokeFromFileAsync<Result>("webSocketServer.js", "start", args: new[] { "success" });

            Assert.AreEqual("success", result?.Message);
            //Debug.Log(result?.Message);

            Result messageResult = await nodeJSService.InvokeFromFileAsync<Result>("webSocketServer.js", "broadcast", args: new[] { "Welcome!" });
            //Debug.Log($"Send : {messageResult?.Message}");


            //Result promiseResult = await nodeJSService.InvokeFromFileAsync<Result>("webSocketServer.js", "exampleMethod", args: new[] { "1" });
            //Debug.Log("Returned promise result");
            //Debug.Log($"Promise result : {promiseResult?.Message}");

        }

        private async void StopServerWithNodeJS()
        {
            await nodeJSService.InvokeFromFileAsync<Result>("webSocketServer.js", "stop", args: new[] { "" });
        }

        public async void BroadcastWebSocketMessage(string message)
        {
            Result messageResult = await nodeJSService.InvokeFromFileAsync<Result>("webSocketServer.js", "broadcast", args: new[] { message });
            Debug.Log($"Send : {messageResult?.Message}");
        }
    }
}

File: Assets/Scripts/Connection/WebSocket/NodeJS/webSocketServer.js
/**
 * Web socket server code that can be used to provide data to the web app.
 */
const WebSocket = require('ws');

/**
 * Web socket server class the use for testing the web app.
 */
class WebSocketServer {
    /** Connected connections for this web socket server. */
    connectedConnection;

    /** Web socket server */
    webSocketServer;

    constructor() {
        this.connectedConnection = new Set();
        this.webSocketServer;
    }

    /**
     * Start the web socket server.
     */
    start(port) {
        this.stop(); // In case we already have a websocket sever running.
        
        debugger;

        console.log(`Starting webSocketServer at port ${port}`);
        this.webSocketServer = new WebSocket.Server({ port: port });
        
        // Add reference to self to use in functions.
        let self = this;

        // Add connection listener that stores and handles the connection.
        this.webSocketServer.on('connection', function connection(webSocket) {
            self.connectedConnection.add(webSocket);
            console.log('Client connected');

            // Echo
            // webSocket.on('message', function incoming(message) {
            //     webSocket.send(`${message}`);
            // });
        
            webSocket.on('close', function () {
                console.log('Client disconnected');
                self.connectedConnection.delete(webSocket)
            });
        });
    }

    /**
     * Stop the web socket server.
     */
    stop() {
        debugger;

        console.log('Closing webSocketServer, if it has not been closed already.')
        this.webSocketServer?.close();
    }
    
    /**
     * Broadcase a message to all connections to the web socket.
     * 
     * @returns The number of connections the message was send to.
     */
    broadcast(message) {
        debugger;

        console.log(`Broadcasting message '${message}' to ${this.connectedConnection.size} connection(s).`);
        this.connectedConnection.forEach(connectedWebSocket => {
            connectedWebSocket.send(message);
        });
        return this.connectedConnection.size;
    }
}

// Setup the webSocketServer.
let webSocketServer = new WebSocketServer();

module.exports = {
    start: async (port) => {
        try {
            webSocketServer.start(parseInt(port));
        } catch(error) {
            return `Error starting web socket server at port ${port}\n` + error.message;
        }
        return `Succesfully started web socket server at port ${port}.`;
    },
    stop: async () => {
        webSocketServer.stop();
    },
    broadcast: async (message) => {
        let connections = webSocketServer.broadcast(message);
        return `Broadcasted message to ${connections} connections.`;
    },
}